### ISSUE TITLE
Accuracy with ModelNet40

### ISSUE THREAD

**Issue by dogyoonlee** [HuguesTHOMAS/KPConv-PyTorch#38]
Hello

I really appreciate your great work.

I asked this question on KPConv with Tensorflow implementation previously

So I ask again here. 

However, I have questions related to **classification accuracy** on ModelNet40 with your default setting

All the things I revised are just the follows.

It's just the saving best accuracy, and best_vote accuracy process to compare the other works.

**_in trainer.py,_**

**1) validation part --> to save best accuracy**

revised self.validation like below and appended saving code

------------------------------------------
val_acc, vote_acc = self.validation(net, val_loader, config)

            if best_acc < val_acc:
                best_acc = val_acc
                conv_epoch = self.epoch
                if config.saving:
                    # Get current state dict
                    save_dict = {'epoch': self.epoch,
                                'model_state_dict': net.state_dict(),
                                'optimizer_state_dict': self.optimizer.state_dict(),
                                'saving_path': config.saving_path}

                    # Save current state of the network (for restoring purposes)
                    checkpoint_path = join(checkpoint_directory, 'best_acc_chkp.tar')
                    torch.save(save_dict, checkpoint_path)

            if best_vote_acc < vote_acc:
                best_vote_acc = vote_acc
                conv_epoch_vote = self.epoch
                if config.saving:
                    # Get current state dict
                    save_dict = {'epoch': self.epoch,
                                'model_state_dict': net.state_dict(),
                                'optimizer_state_dict': self.optimizer.state_dict(),
                                'saving_path': config.saving_path}

                    # Save current state of the network (for restoring purposes)
                    checkpoint_path = join(checkpoint_directory, 'best_vote_acc_chkp.tar')
                    torch.save(save_dict, checkpoint_path)
            
            print('>>> epoch: {:d} \nbest_acc: {:.3f} | conv_epoch: {:d} \nbest_vote_acc: {:.3f} | conv_epoch_vote: {:d} \n'.format(self.epoch, 
                                                                                best_acc, conv_epoch, 
                                                                                best_vote_acc, conv_epoch_vote))
                    
            with open(join(config.saving_path, 'best_results.txt'), "a") as file:
                message = '>>> epoch: {:d} \nbest_acc: {:.3f} | conv_epoch: {:d} \nbest_vote_acc: {:.3f} | conv_epoch_vote: {:d} \n'
                file.write(message.format(self.epoch,
                                            best_acc,
                                            conv_epoch,
                                            best_vote_acc,
                                            conv_epoch_vote))

------------------------------------------------------------

**2) revise validation function to return val_acc and vote_acc**
-------------------------------------------------
    def validation(self, net, val_loader, config: Config):

        if config.dataset_task == 'classification':
            _, val_acc, vote_acc = self.object_classification_validation(net, val_loader, config)
        elif config.dataset_task == 'segmentation':
            self.object_segmentation_validation(net, val_loader, config)
        elif config.dataset_task == 'cloud_segmentation':
            self.cloud_segmentation_validation(net, val_loader, config)
        elif config.dataset_task == 'slam_segmentation':
            self.slam_segmentation_validation(net, val_loader, config)
        else:
            raise ValueError('No validation method implemented for this network type')
        return val_acc, vote_acc
--------------------------------------------------------

**3) revise object_classification_validation function to return 3 values, C1, val_acc, vote_acc to use in validation function**
----------------------------------------------------------
                return C1, val_ACC, vote_ACC
------------------------------------------------------------

The code worked and experiment was successfully done.

The best validation accuracy was **93.75** and best vote_accuracy was **92.385** during the training

however, when I test the saved model of above training with test_models.py, 
the best test accuracy was **92.1** max

My questions is that which one is right results on your code?

as I check the data loading part of your code, 2 code use the same test_dataset which is used as val_dataset in training.
Also, the test process between 2 code is regarded almost same as I think.

what makes the accuracy different between 2 codes?
Did I experiment wrong?

**Comment by HuguesTHOMAS** [comment]
Hi @dogyoonlee,

Thanks for your interest. I noticed the same thing in my experiments too. The validation score is usually higher than the test score although they are computed in the same way. I did not have time to look into that a lot so I am not sure what is at stake here. 

One supposition was that the validation vote is including predictions form early stage of training when the network has not converged, and there could be a positive effect on the score in the same way as if we averaged the prediction from multiple different networks. However I don't think that this is the reason, because the validation average is a low pass filter, which means the vote results depend less and less on the early network, but it does not decrease.

I did not find another reason, but if you have an idea, don't hesitate to post it here, so that we can discuss of it.

Best,
Hugues

**Comment by dogyoonlee** [comment]
Hi @HuguesTHOMAS,
I really appreciate your kind reply!

However, I wonder why validation vote have to be performed.
Because, I think the **validation should be processed without any knowledge** given before.

Also, I'm really confused with your validation strategy during training and testing too.

why the _**self.val_probs**_ (in object_classification_validation function) during training and _**self.test_probs**_(in classification_test function) is needed to compute the accuracy?

As I know, validation process is just processed with test data and label without any knowledge.

I'll really appreciate if you give me the answer on this question.

Best,
Dogyoon

**Comment by HuguesTHOMAS** [comment]
In standard machine learning, **validation** should be performed on a subset of the training set and **test** should be performed on a part of the dataset where labels are not available. In ModelNet40, the labels are available for all objects and the authors just decided to separate the data in a training and test set. In this case this test set is similar to a validation set. This is why for this dataset, validation and test are the same in my code.

Concerning the voting strategy, this is a very common thing to do when using data augmentation. If you test your network on a shape, you can test it several times with different orientations and average the predicted probabilities to get a better and more consistent score. This is what I do with `self.val_probs` and `self.test_probs`, I do not add any knowledge, I just averaged predictions over different orientations (and scales) of the input shapes.

**Comment by dogyoonlee** [comment]
@HuguesTHOMAS 

I really appreciate to your kind and fast reply!!

I understand your explanation about the voting strategy.
As I understand, I think the voting strategy can be regarded as test ensemble technique.

For another issue, which is about the difference between validation accuracy and test accuracy despite of almost same structures of two datasets, I still have a curiosity about the difference.

If I find another possible reason, I will post you again.

Thank you for your reply again.

Best,
Dogyoon

**Comment by densechen** [comment]
@HuguesTHOMAS @dogyoonlee 
I run the training and testing code of ModelNet40 without any modification of default parameters, however, I got a quiet low accuracy, only ~91.4%.

And I use a TITAN XP Card, any wrong?

```bash
Test Accuracy = 91.4%
Test vote 96 : 0.5% (timings : 109.83 121.46)
Test vote 96 : 4.5% (timings : 72.96 124.18)
Test vote 96 : 8.5% (timings : 48.50 126.19)
Test vote 96 : 12.5% (timings : 32.34 129.76)
Test vote 96 : 16.5% (timings : 21.57 128.08)
Test vote 96 : 20.5% (timings : 14.41 128.13)
Test vote 96 : 24.5% (timings : 9.66 127.26)
Test vote 96 : 29.0% (timings : 6.19 126.03)
Test vote 96 : 33.5% (timings : 4.07 124.70)
Test vote 96 : 38.0% (timings : 2.77 122.76)
Test vote 96 : 42.5% (timings : 1.89 121.26)
Test vote 96 : 46.5% (timings : 1.59 122.61)
Test vote 96 : 51.0% (timings : 1.15 121.94)
Test vote 96 : 55.0% (timings : 1.02 125.71)
Test vote 96 : 59.5% (timings : 0.83 123.52)
Test vote 96 : 64.0% (timings : 0.65 120.72)
Test vote 96 : 68.5% (timings : 0.52 119.09)
Test vote 96 : 73.0% (timings : 0.43 117.13)
Test vote 96 : 77.5% (timings : 0.37 116.46)
Test vote 96 : 82.0% (timings : 0.34 115.86)
Test vote 96 : 87.0% (timings : 0.29 113.78)
Test vote 96 : 92.0% (timings : 0.28 111.66)
Test vote 96 : 96.5% (timings : 0.25 111.76)
Test vote 96 : 101.0% (timings : 0.22 112.03)
Test Accuracy = 91.4%
Test vote 97 : 0.5% (timings : 124.35 111.54)
Test vote 97 : 4.5% (timings : 82.63 117.58)
Test vote 97 : 8.5% (timings : 54.94 121.43)
Test vote 97 : 12.5% (timings : 36.55 124.11)
Test vote 97 : 16.5% (timings : 24.35 126.46)
Test vote 97 : 20.5% (timings : 16.28 125.52)
Test vote 97 : 25.0% (timings : 10.35 123.43)
Test vote 97 : 29.5% (timings : 6.63 121.59)
Test vote 97 : 34.0% (timings : 4.31 120.46)
Test vote 97 : 38.5% (timings : 2.83 120.20)
Test vote 97 : 42.5% (timings : 2.00 123.02)
Test vote 97 : 46.5% (timings : 1.44 125.91)
Test vote 97 : 51.0% (timings : 1.04 124.27)
Test vote 97 : 55.0% (timings : 0.93 125.47)
Test vote 97 : 59.0% (timings : 0.73 126.34)
Test vote 97 : 63.5% (timings : 0.56 122.42)
Test vote 97 : 68.0% (timings : 0.44 122.05)
Test vote 97 : 72.5% (timings : 0.36 120.22)
Test vote 97 : 77.0% (timings : 0.41 120.52)
Test vote 97 : 81.5% (timings : 0.46 120.26)
Test vote 97 : 86.0% (timings : 0.46 120.08)
Test vote 97 : 90.5% (timings : 0.37 117.90)
Test vote 97 : 95.0% (timings : 0.31 117.83)
Test vote 97 : 99.5% (timings : 0.25 115.35)
Test Accuracy = 91.5%
Test vote 98 : 0.5% (timings : 73.37 115.06)
Test vote 98 : 1.5% (timings : 106.78 114.29)
Test vote 98 : 6.0% (timings : 67.42 114.71)
Test vote 98 : 10.5% (timings : 42.63 114.59)
Test vote 98 : 15.0% (timings : 27.01 113.89)
Test vote 98 : 19.5% (timings : 17.14 115.47)
Test vote 98 : 24.0% (timings : 10.90 117.32)
Test vote 98 : 28.5% (timings : 7.00 119.77)
Test vote 98 : 32.5% (timings : 4.78 122.63)
Test vote 98 : 37.0% (timings : 3.18 122.23)
Test vote 98 : 41.5% (timings : 2.14 120.15)
Test vote 98 : 46.0% (timings : 1.44 119.11)
Test vote 98 : 50.0% (timings : 1.20 122.47)
Test vote 98 : 54.0% (timings : 0.91 123.94)
Test vote 98 : 58.5% (timings : 0.67 121.81)
Test vote 98 : 63.0% (timings : 0.54 120.17)
Test vote 98 : 67.0% (timings : 0.55 124.34)
Test vote 98 : 71.5% (timings : 0.45 122.02)
Test vote 98 : 76.0% (timings : 0.37 120.66)
Test vote 98 : 80.0% (timings : 0.44 123.53)
Test vote 98 : 84.5% (timings : 0.39 122.22)
Test vote 98 : 89.0% (timings : 0.34 122.05)
Test vote 98 : 93.0% (timings : 0.33 124.61)
Test vote 98 : 97.5% (timings : 0.28 121.78)
Test Accuracy = 91.5%
Test vote 98 : 0.5% (timings : 117.39 119.59)
Test vote 98 : 5.0% (timings : 74.09 117.90)
Test vote 98 : 10.0% (timings : 44.47 115.29)
Test vote 98 : 12.5% (timings : 34.72 138.37)
Test vote 98 : 16.5% (timings : 23.30 134.49)
Test vote 98 : 21.0% (timings : 14.83 127.81)
Test vote 98 : 25.5% (timings : 9.47 123.64)
Test vote 98 : 30.0% (timings : 6.08 119.70)
Test vote 98 : 34.5% (timings : 4.01 118.81)
Test vote 98 : 39.5% (timings : 2.54 115.58)
Test vote 98 : 44.0% (timings : 1.71 115.44)
Test vote 98 : 48.5% (timings : 1.20 115.56)
Test vote 98 : 53.0% (timings : 0.84 115.89)
Test vote 98 : 57.5% (timings : 0.63 115.64)
Test vote 98 : 62.0% (timings : 0.52 116.02)
Test vote 98 : 66.0% (timings : 0.65 118.95)
Test vote 98 : 70.5% (timings : 0.60 119.04)
Test vote 98 : 75.0% (timings : 0.47 117.19)
Test vote 98 : 79.5% (timings : 0.43 118.09)
Test vote 98 : 84.0% (timings : 0.38 116.51)
Test vote 98 : 88.5% (timings : 0.36 118.92)
Test vote 98 : 93.0% (timings : 0.31 120.30)
Test vote 98 : 97.5% (timings : 0.25 119.65)
Test Accuracy = 91.4%
Test vote 99 : 0.5% (timings : 92.25 117.60)
Test vote 99 : 3.5% (timings : 79.06 120.33)
Test vote 99 : 7.0% (timings : 63.63 122.70)
Test vote 99 : 11.0% (timings : 42.31 126.40)
Test vote 99 : 15.0% (timings : 28.22 127.86)
Test vote 99 : 19.0% (timings : 18.84 127.16)
Test vote 99 : 23.0% (timings : 12.66 128.38)
Test vote 99 : 27.5% (timings : 8.14 126.63)
Test vote 99 : 31.5% (timings : 5.53 127.69)
Test vote 99 : 35.5% (timings : 3.83 127.94)
Test vote 99 : 40.0% (timings : 2.59 123.37)
Test vote 99 : 44.5% (timings : 1.77 120.74)
Test vote 99 : 49.0% (timings : 1.22 119.73)
Test vote 99 : 53.5% (timings : 0.84 118.53)
Test vote 99 : 58.0% (timings : 0.67 119.89)
Test vote 99 : 62.5% (timings : 0.53 117.51)
Test vote 99 : 67.0% (timings : 0.43 117.54)
Test vote 99 : 71.5% (timings : 0.41 117.32)
Test vote 99 : 75.5% (timings : 0.46 121.33)
Test vote 99 : 79.5% (timings : 0.58 126.64)
Test vote 99 : 83.5% (timings : 0.48 127.34)
Test vote 99 : 87.5% (timings : 0.47 126.44)
Test vote 99 : 91.5% (timings : 0.42 125.75)
Test vote 99 : 96.0% (timings : 0.33 123.93)
Test vote 99 : 100.5% (timings : 0.26 122.50)
Test Accuracy = 91.4%


```
